### PR TITLE
fix(tcp): harden TCP transport teardown, send paths, and thread ownership

### PR DESCRIPTION
**Pull request by ruck314**
### Description
Hardens the TCP memory and stream bridge implementations so teardown and ZeroMQ multipart send failures are handled explicitly instead of relying on unchecked or partially torn-down state.

### Details
This updates `memory::TcpClient`, `memory::TcpServer`, and `stream::TcpCore` to use RAII-owned worker threads, nullptr-safe socket/context teardown, and deprecated `close()` wrappers that steer new code toward `stop()`.

`memory::TcpServer` and `stream::TcpCore` now check multipart send failures and rebuild the affected PUSH socket to clear ZeroMQ multipart state before later sends. `TcpCore::stop()` also serializes teardown with in-flight `acceptFrame()` calls, and `memory::TcpClient` drops late responses for expired transactions with an explicit warning.

The Python lifecycle surface is updated so managed code can call `_stop()` on the TCP bridge classes. The memory and stream TCP bridge docs now describe the preferred lifecycle calls and the legacy `close()` behavior.

A protected `memory::TcpServer::sendResponseMsg_()` hook was added under the internal C++ surface so the functional C++ regression test can inject a real multipart send failure while still exercising the production worker thread and ZeroMQ sockets.

### CODE REVIEW

File: src/rogue/interfaces/memory/TcpServer.cpp [hunk 1]
```diff
@@ -286,8 +284,29 @@ void rim::TcpServer::runThread() {
...
+                    // Close any remaining parts that were never sent.
+                    for (uint32_t y = x + 1; y < 6; y++) zmq_msg_close(&(msg[y]));
+                    break;
+                }
+            }
+            if (sendFailed) {
+                bridgeLog_->error("Aborted multi-part reply for id=%" PRIu32
+                                  "; remote will see no response for this transaction",
+                                  id);
+            }
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/TcpServer.cpp]: Valid finding — addressed in 168e1cedc.

You are right that the previous "abort and close remaining parts" path only avoided the partial reply if the failure happened on part 0. Once part 0..k-1 had been queued (or flushed) on the PUSH socket, the peer could observe a torso-only reply and the local socket's multipart FSM was stuck expecting the remainder of the failed message — any subsequent transaction reply would have been appended to it.

On send failure we now close and rebind `zmqResp_` to reset the FSM (PULL peers reconnect transparently). If the rebuild itself fails, `runThread` returns without clearing `threadEn_`, so the destructor's `stop()` still joins and tears down the bridge cleanly. The log line is updated to say the peer may have observed a partial reply rather than no reply.

File: src/rogue/interfaces/memory/TcpClient.cpp [hunk 2]
```diff
@@ -411,9 +409,14 @@ void rim::TcpClient::runThread() {
...
 
-            // Transaction expired
+            // Transaction expired — complete with error so the caller is
+            // unblocked.  Without this, the transaction stays in the in-flight
+            // map and the originating doTransaction() blocks until its own
+            // timeout expires, which can be much longer than the bridge
+            // timeout that already fired.
             if (tran->expired()) {
                 bridgeLog_->warning("Transaction expired. Id=%" PRIu32, id);
+                tran->error("TcpClient transaction timeout: Id=%" PRIu32, id);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/TcpClient.cpp]: Valid finding — addressed in 168e1cedc.

You are right on all three points:

1. `Transaction::expired()` returns true when `done_` is set or `iter_` has been cleared, both of which happen when `wait()` returns (timeout path sets `done_=true`, then the post-loop reset clears `iter_`).
2. `Slave::getTransaction(id)` already erases the entry from `tranMap_` before this branch runs, so the in-flight map cleanup the previous comment claimed was never the issue.
3. Calling `tran->error()` on a transaction whose `done_` is already true just appends to `error_` on a completed object — `wait()` had already returned the per-transaction timeout to the caller, and nothing reads the mutated state afterwards.

I removed the `tran->error()` call and updated the comment/log to make it clear this is a stale-response drop, matching the long-standing pattern in `protocols/srp/SrpV0.cpp`. The corresponding TcpServer-side fix (closing+rebinding the PUSH socket on partial multipart failure) lives in the same commit.

File: src/rogue/interfaces/memory/TcpServer.cpp [hunk 3]
```diff
@@ -286,8 +284,78 @@ void rim::TcpServer::runThread() {
...
+                                          this->respAddr_.c_str(), zmq_strerror(zmq_errno()));
+                        rebuilt = false;
+                    }
+                }
+
+                if (!rebuilt) {
+                    if (this->zmqResp_ != nullptr) {
+                        zmq_close(this->zmqResp_);
+                        this->zmqResp_ = nullptr;
+                    }
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/TcpServer.cpp]: Valid finding — addressed in de58893e5.

You're right that the rebuild-failure path nulls `zmqResp_` and returns from `runThread()` while `threadEn_` is still true, which left `stop()`/the dtor calling `zmq_close(this->zmqResp_)` unconditionally. Modern libzmq tolerates a NULL handle (returns -ENOTSOCK), but that's not guaranteed across versions and it's the wrong shape regardless.

Went with option (a): every `zmq_close` / `zmq_ctx_destroy` call in `stop()` is now guarded by a nullptr check (and clears the member after close), matching the pattern already used in the constructor's catch block. Teardown is now safe whether the worker exited via the normal `threadEn_=false` loop or the rebuild-failure early return.

File: src/rogue/interfaces/memory/TcpServer.cpp [hunk 4]
```diff
@@ -286,8 +284,78 @@ void rim::TcpServer::runThread() {
...
+                    zmq_close(this->zmqResp_);
+                    this->zmqResp_ = nullptr;
+                }
+
+                this->zmqResp_ = zmq_socket(this->zmqCtx_, ZMQ_PUSH);
+                bool rebuilt = (this->zmqResp_ != nullptr);
+
+                if (rebuilt) {
+                    uint32_t lopt = 0;
+                    if (zmq_setsockopt(this->zmqResp_, ZMQ_LINGER, &lopt, sizeof(int32_t)) != 0) {
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/TcpServer.cpp]: Valid finding — addressed in de58893e5.

You're right: `ZMQ_LINGER` is documented as taking an `int`, and declaring `lopt` as `uint32_t` while passing `sizeof(int32_t)` was a signedness/type mismatch even though the two sizes happen to coincide on every supported platform.

Switched the rebuild-path variable to `int32_t lopt = 0` and changed the length argument to `sizeof(lopt)` so the variable type and option-length agree at the call site. Limited the change to the new code added by this PR; the constructor uses the same `uint32_t opt` pattern but is pre-existing and outside the scope of this PR.

File: tests/integration/test_memory_tcp_lifecycle_audit_repros.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+# These are integration-tier tests (they test transport-bound code paths)
+# and are tagged @pytest.mark.integration per REP-03.  The source-text
+# invariant pattern is fully deterministic.
+
+import pathlib
+
+import pytest
+
+# Locate the repository root relative to this file:
+#   tests/integration/<this file>  -> 3 levels up -> repo root
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/integration/test_memory_tcp_lifecycle_audit_repros.py]: Comment clarified in 6cbfc9248.

The original "3 levels up" was technically defensible — `parents[N]` is the (N+1)-th ancestor, so `parents[2]` is the 3rd ancestor (3 hops from the file: file → tests/integration → tests → repo root). However, the same phrasing also reads naturally as "3 hops from the `tests/integration` directory", which is wrong, so the comment was ambiguous.

Replaced it with an explicit index-by-index mapping that cannot be misread:

```python
# Locate the repository root.  pathlib.Path.parents indexes from 0,
# so parents[0] = tests/integration, parents[1] = tests,
# parents[2] = repository root.
_REPO_ROOT = pathlib.Path(__file__).resolve().parents[2]
```

File: include/rogue/interfaces/memory/TcpServer.h [hunk 6]
```diff
@@ -69,7 +69,7 @@ class TcpServer : public rogue::interfaces::memory::Master {
 
     //! \cond INTERNAL
   protected:
-    std::thread* thread_ = nullptr;
+    std::unique_ptr<std::thread> thread_;
     std::atomic<bool> threadEn_{false};
     //! \endcond
```
[the remark this answers is not in the document]
Comment (ruck314) [reply include/rogue/interfaces/memory/TcpServer.h:82]: Valid finding — addressed in 6cbfc9248.

You're right: `memory::TcpClient::close()` and `stream::TcpCore::close()` were both annotated `[[deprecated("Use stop() instead")]]` in this PR, but `memory::TcpServer::close()` was left out and still relied on the doc comment alone, so callers got no compile-time warning.

Added the same attribute on the TcpServer header declaration and added a parallel source-text regression test (`tests/cpp/memory/test_tcpserver_close_deprecated_audit_repro.cpp`) mirroring the existing TcpClient/TcpCore guards so the three transport classes stay aligned for future regressions.

File: include/rogue/interfaces/stream/TcpCore.h [hunk 7]
```diff
@@ -118,8 +118,8 @@ class TcpCore : public rogue::interfaces::stream::Master, public rogue::interfac
     /** @brief Destroys the bridge core and releases resources. */
     ~TcpCore();
 
-    /** @brief Closes active bridge connections. */
-    void close();
+    /** @brief Closes active bridge connections. Deprecated; use stop(). */
+    [[deprecated("Use stop() instead")]] void close();
 
     /** @brief Stops the interface and worker thread. */
     void stop();
```
[the remark this answers is not in the document]
Comment (ruck314) [reply include/rogue/interfaces/stream/TcpCore.h:130]: Valid finding — addressed in 6cbfc9248.

You're right that the PR description claimed RAII thread ownership across all three transport classes but `stream::TcpCore` was still using `std::thread* thread_` with raw `new`/`delete`, leaving the stream side inconsistent with the memory side and exception-unsafe in the same way the memory classes were before this PR.

Migrated the member to `std::unique_ptr<std::thread>`, switched the constructor to `std::make_unique`, and replaced the manual `delete thread_; thread_ = nullptr;` in the ctor catch block and `stop()` with `thread_.reset()`. The existing audit-repro tests for TcpClient/TcpServer cover the same source-text invariant (`new std::thread` must not appear) — the matching TcpCore guard is already in `tests/cpp/stream/test_tcpcore_close_deprecated_audit_repro.cpp`.

[the remark this answers is not in the document]
Comment (ruck314) [reply include/rogue/interfaces/stream/TcpCore.h:130]: Valid finding — addressed in 6cbfc9248.

You're right: the PR description claimed every `zmq_sendmsg` call site was hardened, but `stream::TcpCore::acceptFrame()` still logged a per-part warning and continued sending the rest of the multipart frame. On a PUSH socket that leaves the same multipart-FSM corruption hole the memory `TcpServer` had — once part `k > 0` failed, earlier parts may already have been queued/flushed (peer sees a torso-only frame) and any subsequent send would be appended to the unfinished multipart.

Updated `acceptFrame()` to mirror the new memory `TcpServer` recovery path:

1. Break on first send failure.
2. `zmq_msg_close()` the failed part and every unsent part.
3. `zmq_close()` the broken push socket and rebuild it on the same context with the original socket options (`ZMQ_IMMEDIATE`, `ZMQ_LINGER`).
4. Cached the ctor `server` flag as a member so the rebuild can call the right primitive: `zmq_bind(pushAddr_)` for server mode, `zmq_connect(pushAddr_)` for client mode.
5. If the rebuild itself fails, log and leave `zmqPush_ = nullptr`; subsequent `acceptFrame()` calls will fail their first send and re-attempt the rebuild, and `stop()` is now nullptr-safe across every handle so teardown still works.

`bridgeMtx_` is held across the whole acceptFrame body, so the swap is safe vs. concurrent senders. `zmqPull_` and `runThread()` are independent of `zmqPush_`, so receive-side traffic is unaffected.

File: src/rogue/interfaces/stream/TcpClient.cpp [hunk 8]
```diff
@@ -55,7 +55,8 @@ void ris::TcpClient::setup_python() {
 
     bp::class_<ris::TcpClient, ris::TcpClientPtr, bp::bases<ris::TcpCore>, boost::noncopyable>(
         "TcpClient",
-        bp::init<std::string, uint16_t>());
+        bp::init<std::string, uint16_t>())
+        .def("_stop", &ris::TcpCore::stop);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/stream/TcpClient.cpp:60]: Valid finding — addressed in 6cbfc9248.

You're right: only `stream::TcpClient` and `memory::TcpServer` were exposing `_stop` to Python, while `stream::TcpServer` and `memory::TcpClient` were not. Because `pyrogue.Device._stop()` skips any managed interface that does not have a `_stop` attribute, those two would have leaked their worker threads / ZMQ contexts when used as managed interfaces in a Root tree.

Added the missing `.def("_stop", ...)` to both:

- `src/rogue/interfaces/stream/TcpServer.cpp`: `.def("_stop", &ris::TcpCore::stop)`
- `src/rogue/interfaces/memory/TcpClient.cpp`: `.def("_stop", &rim::TcpClient::stop)`

`_start` is intentionally left only on `memory::TcpClient` (where it gates the readiness probe) and `memory::TcpServer` (managed-lifecycle no-op for symmetry) — the stream classes have no readiness handshake or other managed-startup work, so a `_start` no-op there would be inert. Happy to add inert `_start` shims for full symmetry if you'd prefer.

File: src/rogue/interfaces/stream/TcpCore.cpp [hunk 9]
```diff
@@ -246,14 +255,95 @@ void ris::TcpCore::acceptFrame(ris::FramePtr frame) {
...
+            if (this->zmqPush_ != nullptr) {
+                zmq_close(this->zmqPush_);
+                this->zmqPush_ = nullptr;
+            }
+            bridgeLog_->error("Unable to recover TcpCore push socket on %s; "
+                              "subsequent acceptFrame() calls will fail until teardown",
+                              this->pushAddr_.c_str());
+            // stop()'s teardown is now nullptr-safe on zmqPush_, so leaving
+            // the socket nulled is safe across both normal and error exits.
+        }
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/stream/TcpCore.cpp]: Valid finding — addressed in 71e674307.

You're right: after the rebuild-failure path nulled `zmqPush_`, the very next `acceptFrame()` would have called `zmq_sendmsg(nullptr, ...)`, and libzmq's behavior on that call is implementation-defined and not safe to rely on across versions.

Fix:

1. Extracted the close+create+configure+bind/connect sequence into a private `TcpCore::rebuildPushSocket()` helper so both call sites use the same logic and the post-condition is well-defined: on success `zmqPush_` is valid; on any failure it is left `nullptr` and any partially-configured handle is closed before returning.
2. Guarded the top of `acceptFrame()` (after `bridgeMtx_` is taken) with `if (this->zmqPush_ == nullptr) rebuildPushSocket()`. A transient rebuild failure now self-heals on the next frame instead of stranding the bridge until teardown — and if the retry also fails, the frame is dropped with a warning and we return cleanly without ever invoking `zmq_sendmsg` on a null handle.
3. The send-failure recovery path now delegates to the same helper, so the inline rebuild block (which itself contained the `zmqPush_ = nullptr` assignment that could survive the function) is gone.

Added a source-text audit-repro test (`tests/cpp/stream/test_tcpcore_acceptframe_null_push_audit_repro.cpp`) that asserts `acceptFrame()` contains a nullptr guard on `zmqPush_` before its first `zmq_sendmsg` call site, so this regression cannot reappear silently.

File: src/rogue/interfaces/stream/TcpCore.cpp [hunk 10]
```diff
@@ -193,18 +194,83 @@ void ris::TcpCore::stop() {
...
+            zmq_close(zmqPull_);
+            zmqPull_ = nullptr;
+        }
+        if (zmqPush_ != nullptr) {
+            zmq_close(zmqPush_);
+            zmqPush_ = nullptr;
+        }
+        if (zmqCtx_ != nullptr) {
+            zmq_ctx_destroy(zmqCtx_);
+            zmqCtx_ = nullptr;
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/stream/TcpCore.cpp:213]: Valid finding — addressed in c841bee5c.

You're right: `acceptFrame()` and `rebuildPushSocket()` take `bridgeMtx_` to serialise their access to `zmqPush_`/`zmqCtx_`, but `stop()` previously closed the sockets and destroyed the context without acquiring the same lock. The worker `join()` only serialises us with `runThread()`'s use of `zmqPull_`; frame senders running on other threads (e.g., a Python `_stop` from one thread while another thread is pushing frames into the bridge) could still be mid-`zmq_sendmsg` when `stop()` freed the handles. That's UB on every libzmq version regardless of how forgiving any one build happens to be.

Fix:

1. `stop()` now takes `bridgeMtx_` *after* the worker `join()` and around the `zmq_close`/`zmq_ctx_destroy` sequence. The join still must run without the mutex (the worker doesn't take it, and holding it would have no effect anyway), but the post-join teardown now waits for any in-flight `acceptFrame()` / `rebuildPushSocket()` to finish before freeing the handles.

2. `acceptFrame()` now bails out at the top (after taking `bridgeMtx_`) when `threadEn_` is false. Combined with #1 this gives the invariant: once a caller holds `bridgeMtx_`, either the bridge is still running (`threadEn_` true and the handles are valid) or `stop()` has already finished tearing everything down — in the latter case we drop the frame with a debug log instead of dereferencing freed sockets.

This is the same pattern the memory `TcpServer` rebuild path relies on: the mutex is the single point of serialisation between sender and lifecycle, and `threadEn_` is the post-stop sentinel.

Added a source-text audit-repro test (`tests/cpp/stream/test_tcpcore_stop_serialization_audit_repro.cpp`) covering both invariants:

- `stop()` must contain a `bridgeMtx_` guard before its first `zmq_close`.
- `acceptFrame()` must contain a `!threadEn_` guard before its first `zmq_sendmsg`.

Local verification: `./scripts/run_linters.sh` clean, full `ctest --test-dir build --output-on-failure -L cpp` suite (now 19 tests including the new one) passes, `pytest tests/integration/test_memory_tcp_lifecycle_audit_repros.py` still passes.

File: tests/cpp/memory/test_tcpserver_send_swallow_audit_repro.cpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        }
+        // Acceptable preceding contexts:
+        //   "= zmq_sendmsg("   (return code captured)
+        //   "(zmq_sendmsg("    (used inside an expression / predicate)
+        //   "!zmq_sendmsg("    (boolean negation as predicate)
+        const char prev = (i > 0) ? hay[i - 1] : '\0';
+        const bool checked = (prev == '=' || prev == '(' || prev == '!' ||
+                              prev == '<' || prev == '>' || prev == ',' ||
+                              prev == '?' || prev == ':' || prev == '|' ||
+                              prev == '&');
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/memory/test_tcpserver_send_swallow_audit_repro.cpp]: Valid finding — addressed in 3e899d1ef.

You're right that the heuristic was brittle on two of the three cases you raised:

1. **Comments / strings.** The raw-string scan would happily match `// zmq_sendmsg(...)` or `"zmq_sendmsg(...)"` if those ever appeared in `TcpServer.cpp` — the same pattern already exists in a comment in `stream/TcpCore.cpp:314`, so this isn't hypothetical.
2. **`return zmq_sendmsg(...)`.** Walking back through whitespace from the `z` lands on `n` (last char of `return`), which isn't in the allowed prev-char set, so a clearly-checked propagated return would have been mis-flagged as unchecked.
3. **`auto rc = zmq_sendmsg(...)`.** This case actually *is* accepted today (prev char is `=`), so I left that part alone.

Fix:

- Added `stripCommentsAndStrings()` — a small C/C++ pre-processor that elides `// ...`, `/* ... */`, `"..."`, and `'...'` (preserving newlines, replacing inner chars with spaces). The scan now operates on the stripped text, so any future doc/log mention of `zmq_sendmsg(` cannot trigger a false unchecked failure.
- Extended `everyCallIsChecked()` to also accept a preceding `return` keyword: when the prev non-space char is alphanumeric, walk back through identifier chars and check that the token is exactly `return`. The `noreturn` boundary case (8 chars, not 6) is correctly rejected.
- Kept the whole-translation-unit scan scope. The original test comment explains why this is intentional: a partial fix that wraps one call in an `if()` while leaving five other bare calls in an unrolled loop must still fail the assertion, and a function-narrowed scan would lose that property.

Validated the heuristic against a stand-alone harness exercising 11 cases — bare statements, `if`/`while` predicates, assignments, `return`, single-line/block-comment occurrences, string-literal occurrences, mixed bare-among-checked, and the `noreturn` boundary case. Local `./scripts/run_linters.sh` clean, target test passes (`ctest -R tcpserver-send-swallow`).

File: src/rogue/interfaces/memory/TcpServer.cpp [hunk 12]
```diff
@@ -286,8 +293,81 @@ void rim::TcpServer::runThread() {
             zmq_msg_init_size(&(msg[5]), result.length());
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/TcpServer.cpp]: Valid finding — addressed in 33ffbccd7.

You're right: `zmq_msg_init_size` can fail (ENOMEM) and the code proceeded directly into `zmq_msg_data`/`memcpy` without checking the return value, which is UB if the allocation failed.

Added a check on the return value. On failure we log the error, close msg[0..4] (which are all valid at that point — either received or successfully initialized via earlier `zmq_msg_init_size` calls), and `continue` to the next loop iteration, skipping the send entirely. The failed msg[5] is not closed because its state after a failed `zmq_msg_init_size` is implementation-defined; it was originally `zmq_msg_init`'d at line 203 but the failed call may have left it in an indeterminate state.

In practice this can only trigger under catastrophic memory exhaustion (the allocation is 2+ bytes — at minimum the "OK" string), but the defensive check is consistent with the rest of this PR's hardening of the send path.

File: tests/cpp/stream/test_tcpcore_close_deprecated_audit_repro.cpp [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+ * Company    : SLAC National Accelerator Laboratory
+ * ----------------------------------------------------------------------------
+ * Description:
+ * Regression tests for stream::TcpCore::close() is marked "deprecated"
+ * in a comment at TcpCore.cpp line ~186 but has no [[deprecated]] attribute
+ * in the header; callers cannot detect the deprecation at compile time.
+ *
+ * Source-text invariant test: reads the TcpCore header and asserts that
+ * [[deprecated]] appears near the close() declaration.  On HEAD it is absent.
+ * ----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/stream/test_tcpcore_close_deprecated_audit_repro.cpp]: Valid finding — addressed in 28d23044f.

Reworded the header comment and TEST_CASE name to invariant/past-tense phrasing:

- Header now says "Previously the deprecation was noted only in a source comment … with no attribute on the declaration"
- TEST_CASE renamed from "missing [[deprecated]]" to "has [[deprecated]] attribute"

The description now stays accurate after merge.

File: tests/cpp/memory/test_tcpclient_close_deprecated_audit_repro.cpp [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+ * Company    : SLAC National Accelerator Laboratory
+ * ----------------------------------------------------------------------------
+ * Description:
+ * Regression tests for memory::TcpClient::close() is marked "deprecated"
+ * in a comment at TcpClient.cpp line ~163 but has no [[deprecated]] attribute
+ * in the header; callers cannot detect the deprecation at compile time.
+ *
+ * Source-text invariant test: reads the TcpClient header and asserts that
+ * [[deprecated]] appears near the close() declaration.  On HEAD it is absent.
+ * ----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/memory/test_tcpclient_close_deprecated_audit_repro.cpp]: Valid finding — addressed in 28d23044f.

Reworded the header comment and TEST_CASE name to invariant/past-tense phrasing:

- Header now says "Previously the deprecation was noted only in a source comment … with no attribute on the declaration"
- TEST_CASE renamed from "missing [[deprecated]]" to "has [[deprecated]] attribute"

The description now stays accurate after merge.

File: tests/integration/test_memory_tcp_lifecycle_audit_repros.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+def test_memory_tcpclient_uses_unique_ptr_thread():
+    """memory::TcpClient ctor must NOT use raw ``new std::thread``.
+
+    src/rogue/interfaces/memory/TcpClient.cpp line ~120 does::
+
+        this->thread_ = new std::thread(&rim::TcpClient::runThread, this);
+
+    The fix is to use std::unique_ptr<std::thread>.  Until the fix is applied
+    this assertion fails deterministically.
+    """
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/integration/test_memory_tcp_lifecycle_audit_repros.py]: Valid finding — addressed in 28d23044f.

Reworded the docstrings from "Until the fix is applied this assertion fails" to past-tense phrasing ("Previously … used … This was replaced with … This test guards against regression") so the descriptions remain accurate as regression guards after merge.

File: tests/cpp/memory/test_tcpserver_send_swallow_audit_repro.cpp [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+ * ----------------------------------------------------------------------------
+ * Description:
+ * Regression tests for memory::TcpServer::runThread send loop calls
+ * zmq_sendmsg() six times but ignores the return value on all calls; a
+ * partial multi-part message leaves the PUSH socket in an undefined state
+ * without any error propagation.
+ *
+ * Source-text invariant test: reads TcpServer.cpp and verifies each
+ * zmq_sendmsg call has a return-code check.  On HEAD all 6 calls ignore
+ * the return value.
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/memory/test_tcpserver_send_swallow_audit_repro.cpp]: Valid finding — addressed in 28d23044f.

Reworded the header comment and TEST_CASE name to invariant/past-tense phrasing:

- Header now says "Previously the send loop called zmq_sendmsg() six times ignoring the return value"
- TEST_CASE renamed from "ignores zmq_sendmsg returns" to "checks zmq_sendmsg returns"

The description now stays accurate after merge.

**Comment by bengineerd** [comment]
https://github.com/slaclab/rogue/actions/runs/25396372814

It seems like this creates a rather large performance regression.
But maybe it's just variability in the CI VM load.

File: src/rogue/interfaces/stream/TcpCore.cpp [hunk 17]
```diff
@@ -220,6 +296,35 @@ void ris::TcpCore::acceptFrame(ris::FramePtr frame) {
...
+        if (!rebuildPushSocket()) {
+            bridgeLog_->warning("Push socket unavailable on %s; dropping frame "
+                                "(payload=%" PRIu32 ")",
+                                this->pushAddr_.c_str(),
+                                frame->getPayload());
+            return;
+        }
+    }
+
     if ((zmq_msg_init_size(&(msg[0]), 2) < 0) ||  // Flags
```
Comment (bengineerd) [inline src/rogue/interfaces/stream/TcpCore.cpp]: `acceptFrame()` can leak already-initialized ZMQ message parts on these init failure paths. If `msg[1]` or `msg[2]` fails here, any earlier header parts remain initialized; if `msg[3]` fails below, `msg[0..2]` remain initialized. Please track which `zmq_msg_init_size()` calls succeeded and `zmq_msg_close()` those parts before returning.

**Review by bengineerd — COMMENTED** [review]
One inline cleanup finding from review.

File: src/rogue/interfaces/memory/TcpClient.cpp [hunk 18]
```diff
@@ -462,7 +469,8 @@ void rim::TcpClient::setup_python() {
         bp::init<std::string, uint16_t, bp::optional<bool> >())
         .def("close", &rim::TcpClient::close)
```
Comment (bengineerd) [inline src/rogue/interfaces/memory/TcpClient.cpp:462]: This binding now takes the address of a method declared with `[[deprecated]]`, which itself triggers `-Wdeprecated-declarations` in normal C++ compilation. If the project or a downstream build treats that warning as an error, the Boost.Python binding can fail to compile even though it is intentionally preserving the legacy Python `close()` API. Please bind `close` through a non-deprecated wrapper that calls `stop()`, or locally suppress the deprecation warning around the binding.

**Review by bengineerd — COMMENTED** [review]
One additional inline build-warning finding from review.

File: src/rogue/interfaces/stream/TcpCore.cpp (hunk 17 above)
Comment (ruck314) [reply src/rogue/interfaces/stream/TcpCore.cpp]: Valid finding — addressed in 1af32c085.

The compound `||` expression meant that if `msg[1]` or `msg[2]` init failed, earlier successfully-initialized parts leaked; same for `msg[3]` failing after `msg[0..2]` succeeded.

Fixed by splitting into individual checks, each closing all previously-initialized parts before returning.

File: src/rogue/interfaces/memory/TcpClient.cpp (hunk 18 above)
Comment (ruck314) [reply src/rogue/interfaces/memory/TcpClient.cpp:462]: Valid finding — addressed in 1af32c085.

Taking the address of a `[[deprecated]]` method triggers `-Wdeprecated-declarations`, which breaks builds that use `-Werror`. The binding intentionally preserves the legacy Python `close()` API while the C++ attribute steers new C++ callers toward `stop()`.

Wrapped all three `.def("close", ...)` sites (memory::TcpClient, memory::TcpServer, stream::TcpCore) in `#pragma GCC diagnostic` push/pop to locally suppress the warning. This keeps the Python API stable without polluting the build output.

**Review by bengineerd — APPROVED** [review]
